packages/core/src/modules/oob/messages/OutOfBandMessage.ts

```diff
@@ -0,0 +1,123 @@
+}
+
+export class OutOfBandMessage extends AgentMessage {
+  public constructor(options: OutOfBandMessageOptions) {
+    super()
+    if (options) {
+      this.id = options.id ?? this.generateId()
+      this.label = options.label
+      this.goalCode = options.goalCode
+      this.goal = options.goal
+      this.accept = options.accept
+      this.handshakeProtocols = options.handshakeProtocols
+      this.services = options.services
+    }
+  }
+
+  public addRequest(message: AgentMessage) {
+    if (!this.requests) this.requests = []
+    const requestAttachment = new Attachment({
+      id: this.generateId(),
+      mimeType: 'application/json',
+      data: new AttachmentData({
+        base64: JsonEncoder.toBase64(message.toJSON()),
+      }),
+    })
+    this.requests.push(requestAttachment)
+  }
+
+  public getRequests(): PlaintextMessage[] | undefined {
+    return this.requests?.map((request) => {
+      if (request.data.base64) {
+        return JsonEncoder.fromBase64(request.data.base64)
+      }
+
+      if (request.data.json) {
+        return request.data.json
+      }
+
+      throw new AriesFrameworkError('There is no supported data attribute in attachment')
+    })
+  }
+
+  public toUrl({ domain }: { domain: string }) {
+    const invitationJson = this.toJSON()
+    const encodedInvitation = JsonEncoder.toBase64URL(invitationJson)
+    const invitationUrl = `${domain}?oob=${encodedInvitation}`
+    return invitationUrl
+  }
+
+  public static async fromUrl(invitationUrl: string) {
+    const parsedUrl = parseUrl(invitationUrl).query
+    const encodedInvitation = parsedUrl['oob']
+
+    if (typeof encodedInvitation === 'string') {
+      const invitationJson = JsonEncoder.fromBase64(encodedInvitation)
+      const invitation = this.fromJson(invitationJson)
+
+      return invitation
+    } else {
+      throw new AriesFrameworkError(
+        'InvitationUrl is invalid. It needs to contain one, and only one, of the following parameters; `oob`'
+      )
+    }
+  }
+
+  public static async fromJson(json: Record<string, unknown>) {
+    const invitation = JsonTransformer.fromJSON(json, OutOfBandMessage)
+    await MessageValidator.validate(invitation)
+    return invitation
+  }
+
+  @Equals(OutOfBandMessage.type)
+  public readonly type = OutOfBandMessage.type
+  public static readonly type = `https://didcomm.org/out-of-band/1.1/invitation`
+
+  public readonly label?: string
+
+  @Expose({ name: 'goal_code' })
+  public readonly goalCode?: string
+
+  public readonly goal?: string
+
+  public readonly accept?: string[]
+
+  @Expose({ name: 'handshake_protocols' })
+  public handshakeProtocols?: string[]
+
+  @Expose({ name: 'requests~attach' })
+  @Type(() => Attachment)
+  @IsArray()
+  @ValidateNested({
+    each: true,
+  })
+  @IsInstance(Attachment, { each: true })
+  @IsOptional()
+  private requests?: Attachment[]
+
+  @IsArray()
+  @ArrayNotEmpty()
+  public services: DidCommService[] = []
```

Comment:
Can also be string values. 


```suggestion
  public services: Array<DidCommService | string> = []
```

Reply:
This change required resolving the did if the services contain a string, so I added it.

Reply:
> resolving the did if the services contain a string

I'm not sure if we're quite ready for that. We don't support public dids yet to create connections, and the did resolver returns the new diddoc format while the rest of AFJ is still built on the old format.

We should write some transformers for the connection record diddoc, and update it in the connection reocrd and also integrate did based message sending in the message sender (so we don't store the did document in the connection record, but rather resolve it (or get from cache) when sending a message)

Reply:
Should we throw an error when there is a string in that array? Then, maybe, this change doesn't make sense.